vehicles/Hero/main.cc

```diff
@@ -0,0 +1,828 @@
+  }
+}
+
+//==================================================================================================
+// Gimbal
+//==================================================================================================
+
+const osThreadAttr_t gimbalTaskAttribute = {.name = "gimbalTask",
+                                            .attr_bits = osThreadDetached,
+                                            .cb_mem = nullptr,
+                                            .cb_size = 0,
+                                            .stack_mem = nullptr,
+                                            .stack_size = 512 * 4,
+                                            .priority = (osPriority_t)osPriorityHigh,
+                                            .tz_module = 0,
+                                            .reserved = 0};
+osThreadId_t gimbalTaskHandle;
+// TODO:
+// 1. test keyboard and mouse data
+
+// Params Initialization
+static control::Motor4310* pitch_motor = nullptr;
+//static bsp::Laser* laser = nullptr;
+
+void gimbalTask(void* arg) {
+  UNUSED(arg);
+  // motor pointer initialization
+  control::Motor4310* motors_can2_gimbal[] = {pitch_motor};
+  control::Motor4310* motors_can1_gimbal[] = {yaw_motor};
+
+  print("Wait for beginning signal...\r\n");
+  RGB->Display(display::color_red);
+//  laser->On(); // has some problem.
+  // waiting for start signal
+  while (true) {
+    if (dbus->keyboard.bit.B || dbus->swr == remote::DOWN) break;
+    osDelay(100);
+  }
+
+//  pitch_motor->SetZeroPos(); //（for test）
+//  yaw_motor->SetZeroPos(); // for test
+  osDelay(1000); // wait for 4310s enable
+
+  // the start code for motor 4310s
+  pitch_motor->MotorEnable();
+  yaw_motor->MotorEnable();
+  osDelay(GIMBAL_TASK_DELAY);
+
+  // initialize start position
+  RGB->Display(display::color_yellow);
+//  laser->Off();
+  // need test Pos_Vel mode params
+  pitch_motor->SetOutput(0, 10);
+  // need test MIT mode params
+  yaw_motor->SetOutput(0, 0, 8, 1.5, 0);
+  control::Motor4310::TransmitOutput(motors_can1_gimbal, 1);
+  control::Motor4310::TransmitOutput(motors_can2_gimbal, 1);
+  osDelay(200);
+
+  print("Gimbal Begin!\r\n");
+  RGB->Display(display::color_green);
+//  laser->On();
+
+  float pitch_keyboard = 0; // E is lifting, Q is lowering
+  float pitch_mouse = 0, yaw_mouse = 0;
+  float pitch_remote = 0, yaw_remote = 0;
+  float pitch_sum = 0;
+  float pitch_pos = 0, yaw_pos = 0;
+  while (true) {
+    // Dead
+    while (Dead) osDelay(100);
+
+    // Data Collection
+    // pitch data from keyboard
+    if (dbus->keyboard.bit.E) pitch_keyboard -= PI; // offset need test
+    if (dbus->keyboard.bit.Q) pitch_keyboard += PI; // offset need test
+    if (-0.55 * PI <= pitch_keyboard && pitch_keyboard <= 0.55 * PI) pitch_keyboard = 0;
+    if (pitch_keyboard > 0)
+      pitch_keyboard -= 0.85 * PI;
+    if (pitch_keyboard < 0)
+      pitch_keyboard += 0.85 * PI;
+    pitch_keyboard = clip<float>(pitch_keyboard, -30 * PI, 30 * PI);
```

Comment:
Instead of using all the if conditions. What if we just increase pitch_keyboard 1.5pi when Q is pressed, and decrease 1.5 pi when E is pressed?

Reply:
In this situation, pitch_keyboard is like a discrete angular velocity, rather than angle. Therefore, we want this velocity decrease to 0 when we do not press the keyboard.